Add option to make predictable simulation sequence

cowboyd opened the pull request:
Motivation
-----------

In order to make testing and development easier, we wanted the ability to pass in a uuid to make sure that the created simulation id was stable so that queries we had sitting around would continue to work and sequences of snippets used for testing would not be one-use and obsolete.

Approach
----------

We can accomplish the same goal by just using a predictable sequence of UUIDs. That way, when we run a server, and start a simulation, it always gets the same UUID.

Currently, faker does not support independent instances that each have their own pseudo random number generators so this change in its current form depends on an open PR I have on the faker.js repo making multiple faker instances safe.

We'll need this very soon since each simulation will have its own random seed and its own faker instance.

dagda1 commented on packages/server/src/faker.ts:
```diff
@@ -0,0 +1,14 @@
+import Faker from 'faker/lib';
+
+export function createFaker(seed: number): any {
+  let faker = new Faker({ locales: require('faker/lib/locales') });
```
why do we need to use require here?

cowboyd replied:
We _may_ want to defer to loading locales depending on the simulation, but best to keep it simple for now. I'm just going to import it.

dagda1 commented on packages/server/src/schema/context.ts:
```diff
@@ -1,16 +1,15 @@
+    let { scope, simulations, newid } = this;
 
-    let id = uuid || v4();
+    let id = newid();
```
does this guarantees the same uuids each time if I do:

```ts
for (const i = o; i < 5 ; i++) {
  console.log(newid());
``

cowboyd replied:
Yep. here's a quick test:

```ts
import { createFaker } from './faker';

[1,2,3].forEach(trial => {
  console.log(`trial ${trial}`);
  let faker = createFaker(100);
  [1,2,3,4,5].forEach((i) => {
    console.log(`  ${i}. ${faker.datatype.uuid()}`);
  });
});
```

```
trial 1
  1. 8a4668d2-0212-4a3d-b2d9-1e938281034f
  2. fc221d44-a6bf-4ade-9429-52084ac20997
  3. 96176607-e0fc-408e-a98b-fa2910d318dc
  4. d4140d0f-5ea5-4091-9e5c-2c3a09826991
  5. a523e8f6-9f62-45b3-84b3-d28f0fe95b85
trial 2
  1. 8a4668d2-0212-4a3d-b2d9-1e938281034f
  2. fc221d44-a6bf-4ade-9429-52084ac20997
  3. 96176607-e0fc-408e-a98b-fa2910d318dc
  4. d4140d0f-5ea5-4091-9e5c-2c3a09826991
  5. a523e8f6-9f62-45b3-84b3-d28f0fe95b85
trial 3
  1. 8a4668d2-0212-4a3d-b2d9-1e938281034f
  2. fc221d44-a6bf-4ade-9429-52084ac20997
  3. 96176607-e0fc-408e-a98b-fa2910d318dc
  4. d4140d0f-5ea5-4091-9e5c-2c3a09826991
  5. a523e8f6-9f62-45b3-84b3-d28f0fe95b85
```

dagda1 replied:
nice!

dagda1 reviewed (changes requested):
I think it is a good idea to have the uuid for the seed, I looked at using fast check `fc.uuid` but fast-check definitely had issues running outside of its runner.

Do you think you could add a test that would help for documentation also?  Any test at all would be great

cowboyd commented:
@dagda1 I added tests to make sure the ids were stable, and then also refactored the tests a bit to make them a bit easier to setup a server. the `createTestServer` starts a server and then returns a client for it.

dagda1 reviewed (approved):
Deterministic data is essential and great to see it there from the start.  Faker is a better fit than fast-check on retrospect.

cowboyd commented:
> Deterministic data is essential and great to see it there from the start. Faker is a better fit than fast-check on retrospect.

@dagda1 I tried using fast check to start with but while it seemed much more powerful, it also seemed much more geared towards generating edge-cases that were nevertheless valid. For example making sure your system handles an email addres like `aoeunthoeuh+..aatgxvaagg$$@da-oaeu--eogg.11aof.com` which is really handy if you want to stress-test your system, but is not something you would every see IRL. 

We may be able to use it in the future, and we may be able to use the FC integration that you had in the spike in order to make faker based arbitraries.

One thing that kinda stinks right now is that the faker typings are pretty broken. The author has indicated that he would be amenable to accepting typings https://github.com/Marak/faker.js/issues/1023 So perhaps we can propose a setup similar to the one we have in effection v1 where we have typings tests using `dslint`

dagda1 commented:
>>  tried using fast check to start with but while it seemed much more powerful, it also seemed much more geared towards generating edge-cases that were nevertheless valid

@cowboyd  exactly :) it actually tries to break things also.  There are ways to circumvent that but I think it would be hammering fast-check into the solution rather than it being a good fit.

dagda1 commented:
> One thing that kinda stinks right now is that the faker typings are pretty broken. The author has indicated that he would be amenable to accepting typings [Marak/faker.js#1023](https://github.com/Marak/faker.js/issues/1023)

That is interesting, I must take a look
